lib/auth/accountrecovery_test.go

```diff
@@ -517,6 +518,450 @@ func TestApproveAccountRecovery_WithLock(t *testing.T) {
+
+	cases := []struct {
+		name       string
+		getRequest func() *proto.CompleteAccountRecoveryRequest
+	}{
+		{
+			name: "add new TOTP device",
+			getRequest: func() *proto.CompleteAccountRecoveryRequest {
+				secrets, err := srv.Auth().RotateUserTokenSecrets(context.TODO(), approvedToken.GetName())
+				require.NoError(t, err)
+
+				otpCode, err := totp.GenerateCode(secrets.GetOTPKey(), srv.Clock().Now())
+				require.NoError(t, err)
+
+				return &proto.CompleteAccountRecoveryRequest{
+					NewDeviceName:           "new-otp",
+					RecoveryApprovedTokenID: approvedToken.GetName(),
+					NewAuthnCred: &proto.CompleteAccountRecoveryRequest_NewMFAResponse{NewMFAResponse: &proto.MFARegisterResponse{
+						Response: &proto.MFARegisterResponse_TOTP{TOTP: &proto.TOTPRegisterResponse{Code: otpCode}},
+					}},
+				}
+			},
+		},
+		{
+			name: "add new U2F device",
+			getRequest: func() *proto.CompleteAccountRecoveryRequest {
+				u2fRegResp, _, err := getMockedU2FAndRegisterRes(srv.Auth(), approvedToken.GetName())
+				require.NoError(t, err)
+
+				return &proto.CompleteAccountRecoveryRequest{
+					NewDeviceName:           "new-u2f",
+					RecoveryApprovedTokenID: approvedToken.GetName(),
+					NewAuthnCred: &proto.CompleteAccountRecoveryRequest_NewMFAResponse{NewMFAResponse: &proto.MFARegisterResponse{
+						Response: &proto.MFARegisterResponse_U2F{U2F: u2fRegResp},
+					}},
+				}
+			},
+		},
+	}
+
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			req := c.getRequest()
+
+			// Change authentication.
+			err = srv.Auth().CompleteAccountRecovery(ctx, req)
+			require.NoError(t, err)
+
+			// Test events emitted.
+			event := mockEmitter.LastEvent()
+			require.Equal(t, event.GetType(), events.MFADeviceAddEvent)
+			require.Equal(t, event.GetCode(), events.MFADeviceAddEventCode)
+			require.Equal(t, event.(*apievents.MFADeviceAdd).UserMetadata.User, u.username)
+
+			// Test new device has been added.
+			mfas, err := srv.Auth().Identity.GetMFADevices(ctx, u.username, false)
+			require.NoError(t, err)
+
+			found := false
+			for _, mfa := range mfas {
+				if mfa.GetName() == req.NewDeviceName {
+					found = true
+					break
+				}
+			}
+			require.True(t, found, "MFA device %q not found", req.NewDeviceName)
+		})
+	}
+}
+
+func TestCompleteAccountRecovery_WithErrors(t *testing.T) {
+	srv := newTestTLSServer(t)
+	ctx := context.Background()
+	mockEmitter := &events.MockEmitter{}
+	srv.Auth().emitter = mockEmitter
+
+	defaultModules := modules.GetModules()
+	defer modules.SetModules(defaultModules)
+	modules.SetModules(&testWithCloudModules{})
+
+	u, err := createUserWithSecondFactorAndRecoveryCodes(srv)
+	require.NoError(t, err)
+
+	cases := []struct {
+		name           string
+		expErrMsg      string
+		isDuplicate    bool
+		isBadParameter bool
+		getRequest     func() *proto.CompleteAccountRecoveryRequest
+	}{
+		{
+			name: "invalid token type",
+			getRequest: func() *proto.CompleteAccountRecoveryRequest {
```

Comment:
Are we looking for a particular error in here? If not, consider adding a comment so it's clearer to readers.

```suggestion
			// expectErrMsg not supplied on purpose.
			getRequest: func() *proto.CompleteAccountRecoveryRequest {
```